### PR TITLE
Pull request for issue #63

### PR DESCRIPTION
**Pull request by amihaiemil**
@dmarkov This is the PR for issue #63

### CODE REVIEW

**Comment by amihaiemil** [comment]
@Suseika 

2 tests that I wrote for color disabling are currently failing and are ignored.
Problem is that the text isn't decolored correctly when the color is specified in the conversion pattern. E.g. "[%color-blue{%p}] %color-blue{%m}" ,

For instance, in disableConstantColor(), the text is first formatted with color: 

[**\\u001B[34m**DEBUG**\\u001B[m**] **\\u001B[34m**no color text**\\u001B[m**
This is OK.

Then it is formatted with the system property com.jcabi.log.coloring == false.
I would expect the text to be 
[DEBUG] no color text

but it is: 
[**\\u001B[34m**DEBUG] **\\u001B[34m**no color text

which can't be right, in my opinion. But maybe I'm not doing something right?

**Comment by dmarkov** [comment]
@amihaiemil Many thanks, I will find someone to review it before we merge

**Comment by dmarkov** [comment]
@dmzaytsev could you please review this PR

**Comment by dmzaytsev** [comment]
@amihaiemil the puzzle says
>Intergration testing is required because this can't be done in a unit test:
 - *  ...
 - *  - We can't just pass a parameter to `MulticolorLayout` that determines if
 - *  the layout should color its output, because Log4J requires
 - *  EnhancedPatternLayout to have a 0-arguments constructor.
 - *  - We can't disable coloring with something like
 - *  `MulticolorLayout#setColoring(boolean)`, too, (that would be unit-testable)
 - *  because Log4J won't know how to call that.
 - *  - `System.getProperty(String)` is not mockable for obvious reasons.

Why did you implement unit tests instead of integration tests? It's impossible to test in an unit test as the puzzle author mentioned

**Comment by amihaiemil** [comment]
@dmzaytsev In my opinion the author meant "it's not unit testable" as in you cannot test the decoloring behaviour of ``MulticolorLayout`` in isolation (mock all dependencies), because System.getProperty() is not mockable. That's why it is in integration test and not a true unit test.

So that's what I did, I actually set the system property to ``false`` when needed and put it back to ``true`` when the test finished. 

How else would you test this? I only mocked log4j's logging event. Other option would be to actually log a text in the console, retrieve it from there and check the color? All for the same result.

File: src/test/java/com/jcabi/log/MulticolorLayoutTest.java [hunk 1]
```diff
@@ -167,4 +158,146 @@ public void throwsOnIllegalColorName() throws Exception {
...
+                COLORING_PROPERTY, Boolean.TRUE.toString()
+            );
+        }
+    }
+
+    /**
+     * MulticolorLayout disables constant color
+     * when -Dcom.jcabi.log.coloring=false.
+     * @throws Exception - if something goes wrong.
+     */
```
Comment (dmzaytsev) [inline src/test/java/com/jcabi/log/MulticolorLayoutTest.java]: @amihaiemil you have to create a new puzzle to fix the bug and to enable the tests

**Comment by dmzaytsev** [comment]
@amihaiemil okay , let's keep it as is 
please see one comment above

**Comment by amihaiemil** [comment]
@dmzaytsev travis complains. One build complains about a space in the puzzle I pus, but I am not sure what it means. Can you please have a look? https://travis-ci.org/jcabi/jcabi-log/jobs/106301428

The other complains about failing tests: https://travis-ci.org/jcabi/jcabi-log/jobs/106301429

I saw this discussion: https://github.com/jcabi/jcabi-log/pull/61 
Apparently it's because tests are run by multiple threads simultaneously. I am not sure how to deal with that, to be honest. Whatever I do, I still need to set that system property, don't I?

**Comment by dmzaytsev** [comment]
@amihaiemil it means you have to add a space before the first character at the second line of the puzzle and below. Take a look at [the previous puzzle](https://github.com/jcabi/jcabi-log/pull/73/files#diff-63ee2a33bbfb02149dd17e1ceaf0ed41L45)

>/home/travis/.rvm/gems/ruby-1.9.3-p551/gems/pdd-0.15.4/lib/pdd/source.rb:144:in `rescue in puzzles': ["in /home/travis/build/jcabi/jcabi-log/src/test/java/com/jcabi/log/MulticolorLayoutTest.java", #<PDD::Error: Space expected>] (PDD::Error)

**Comment by dmzaytsev** [comment]
@amihaiemil you can't set a system property only for one test that's why the puzzle author suggested to implement an integration test I guess

> The other complains about failing tests: https://travis-ci.org/jcabi/jcabi-log/jobs/106301429
> 
> I saw this discussion: #61 
> Apparently it's because tests are run by multiple threads simultaneously. I am not sure how to deal with that, to be honest. Whatever I do, I still need to set that system property, don't I?

**Comment by amihaiemil** [comment]
@dmzaytsev PLease have a look. I changesd a bit the implementation, to make it testable. Basically I created a class that wraps the check of the system property and this class is passed as a parameter to ``MulticolorLayout``. An interface is used, and thus in unit tests a mock of this class can be passed to ``MulticolorLayout``
. 
Maybe a disadvantage would be that now there is a constructor which is only used by tests, but on the other hand, we let the user specify his ``TextDecolorant`` (maybe they cannot set that system property and need other way to decide whether or not to decolor the text).

**Comment by amihaiemil** [comment]
@dmzaytsev  I made a little mistake in my previous comment. The users cannot use the parameterized constructor, since log4j doesn't know how to call it (as the author mentioned in the original todo). So the parameterless constructor will always be used and the parameterized one remains for test purposes only.

File: src/main/java/com/jcabi/log/MulticolorLayout.java [hunk 2]
```diff
@@ -137,6 +132,27 @@
...
+     */
+    public MulticolorLayout() {
+        this(new RtTextDecolorant());
+    }
+
+    /**
+     * Constructor that allos to specify the decolorant type.
+     * @param dclrnt The given decolorant.
+     */
+    public MulticolorLayout(final TextDecolorant dclrnt) {
```
Comment (dmzaytsev) [inline src/main/java/com/jcabi/log/MulticolorLayout.java]: @amihaiemil we can make it visible only on the package level if it used only in tests

**Comment by dmzaytsev** [comment]
@amihaiemil please see my comment above

**Comment by amihaiemil** [comment]
@dmzaytsev done.

**Comment by dmzaytsev** [comment]
@amihaiemil thank you

**Comment by dmzaytsev** [comment]
@rultor merge

File: src/main/java/com/jcabi/log/RtTextDecolorant.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    /**
+     * Name of the property that is used to disable log coloring.
+     */
+    private static final String COLORING_PROPERY = "com.jcabi.log.coloring";
+
+    @Override
+    public boolean isColoringEnabled() {
+        return !"false".equals(
+                     System.getProperty(RtTextDecolorant.COLORING_PROPERY)
+                 );
```
Comment (yegor256) [inline src/main/java/com/jcabi/log/RtTextDecolorant.java]: @amihaiemil what happened here with indentation?

**Comment by yegor256** [comment]
@amihaiemil this implementation looks like a huge overkill for a simple problem.. creating three classes just to unit test one simple feature? the original ticket suggest to create an integration test. the tests you created just make the code more complex and less maintainable. I'm against this. @dmzaytsev what do you think?

**Comment by dmzaytsev** [comment]
@yegor256 actually  these three classes are very simple classes. Surely  `RtTextDecolorant` and `MkRtTextDecolorant` can be embedded into `TextDecolorant`. So we have only one new class.
Usually unit tests faster and cheaper than integration tests
For integration tests we should start separate process twice for `com.jcabi.log.coloring=false` and `true'
In case of unit tests all tests running in parallel
that's why I accepted this PR

**Comment by yegor256** [comment]
@dmzaytsev yeah, you're right, but in this case, the test is so complicated and the design is so heavy... we won't be able to understand what all this is for, in a few months

**Comment by amihaiemil** [comment]
@yegor256 I fixed the indentation. I made the change in implementation so it can be testable. If  we leave the implementation as it was then the only way to make it testable is to disable the multi threaded  run of the tests.

Also, I'm not sure how to make the tests "skinnier", they are already straight-forward. The reason of the implementation, is explained in the Javadoc of ``TextDecolorant`` .

What should I do?

**Comment by yegor256** [comment]
@amihaiemil I think you should create an integration test (using maven-invoker-plugin) and delete this unit test and this class you created

**Comment by amihaiemil** [comment]
@yegor256 I removed the new classes and wrote the integratin tests.

**Comment by amihaiemil** [comment]
@yegor256 Please give feedback as soon as you get the chance so we can advance or merge and close this PR.

**Comment by amihaiemil** [comment]
@yegor256 Please don't forget to give feedback here. Thanks

**Comment by amihaiemil** [comment]
@yegor256 We need to close this in the next 48 hours, can you let rultor merge or let me know if you think other changes are necessary?

**Comment by amihaiemil** [comment]
@yegor256  What about this one? Can you merge and deploy it if it's ok? It should be closed by tomorrow.

**Comment by dmzaytsev** [comment]
@amihaiemil I would suggest you ask @dmarkov for more time  in the ticket #63

**Comment by amihaiemil** [comment]
@dmzaytsev I already explained the situation and asked Dan if we can wait for Yegor, but he did not reply.

**Comment by amihaiemil** [comment]
@yegor256 ping

**Comment by amihaiemil** [comment]
@yegor256 This is **still** up for merging...

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by amihaiemil** [comment]
@yegor256 Can you deploy this pls? Thanks.

**Comment by yegor256** [comment]
@rultor deploy

**Comment by dmarkov** [comment]
@dmzaytsev Thanks a lot, I just topped your account for **26 mins**, transaction ID `78697047`, total time was 527 hours and 59 mins; you're getting extra minutes here (c=11); +26 to your rating, your total score is [+3675](http://www.netbout.com/b/35166?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy

**Pull request merged**